astropy/utils/data.py

```diff
@@ -1002,9 +1016,90 @@ def connect_ftp(self, user, passwd, host, port, dirs, timeout):
 
 
+@functools.lru_cache()
+def _build_urlopener(ftp_tls=False, ssl_context=None, allow_insecure=False):
+    """
+    Helper for building a `urllib.request.build_opener` which handles TLS/SSL.
+    """
+
+    ssl_context = dict(it for it in ssl_context) if ssl_context else {}
+    cert_chain = {}
+    if 'certfile' in ssl_context:
+        cert_chain.update({
+            'certfile': ssl_context.pop('certfile'),
+            'keyfile': ssl_context.pop('keyfile', None),
+            'password': ssl_context.pop('password', None)
+        })
+    elif 'password' in ssl_context or 'keyfile' in ssl_context:
+        raise ValueError(
+            "passing 'keyfile' or 'password' in the ssl_context argument "
+            "requires passing 'certfile' as well")
+
+    if 'cafile' not in ssl_context and certifi is not None:
+        ssl_context['cafile'] = certifi.where()
+
+    ssl_context = ssl.create_default_context(**ssl_context)
+
+    if allow_insecure:
+        ssl_context.check_hostname = False
+        ssl_context.verify_mode = ssl.CERT_NONE
+
+    if cert_chain:
+        ssl_context.load_cert_chain(**cert_chain)
+
+    https_handler = urllib.request.HTTPSHandler(context=ssl_context)
+
+    if ftp_tls:
+        urlopener = urllib.request.build_opener(_FTPTLSHandler(), https_handler)
+    else:
+        urlopener = urllib.request.build_opener(https_handler)
+
+    return urlopener
+
+
+def _try_url_open(source_url, timeout=None, http_headers=None, ftp_tls=False,
+                  ssl_context=None, allow_insecure=False):
+    """Helper for opening a URL while handling TLS/SSL verification issues."""
+
+    # Always try first with a secure connection
+    # _build_urlopener uses lru_cache, so the ssl_context argument must be
+    # converted to a hashshable type (a set of 2-tuples)
+    ssl_context = frozenset(ssl_context.items() if ssl_context else [])
+    urlopener = _build_urlopener(ftp_tls=ftp_tls, ssl_context=ssl_context,
+                                 allow_insecure=False)
+    req = urllib.request.Request(source_url, headers=http_headers)
+
+    try:
+        return urlopener.open(req, timeout=timeout)
+    except urllib.error.URLError as exc:
+        reason = exc.reason
+        if (isinstance(reason, ssl.SSLError)
+                and reason.reason == 'CERTIFICATE_VERIFY_FAILED'):
+            msg = (f'verification of TLS/SSL certificate at {source_url} '
+                   f'failed: this can mean either the server is '
+                   f'misconfigured or your local root CA certificates are '
+                   f'out-of-date; in the latter case this can usually be '
+                   f'addressed by installing the Python package "certifi" '
+                   f'(see the documentation for download_url)')
+            if not allow_insecure:
+                msg += (f' or in both cases you can work around this by '
+                        f'passing allow_insecure=True; the original error '
```

Comment:
Could polish the wording but I feel like we need to emphasize that one should not do this blindly.
```suggestion
                        f'passing allow_insecure=True but only if you understand the implications; the original error '
```